========================================
TEST SUITE ``100-lazy_matrix_mul``
========================================

Using ``lazy_lazy_matrix_mul``
---------------------

Importing the function from the module:
    >>> lazy_matrix_mul = __import__('101-lazy_matrix_mul').lazy_matrix_mul
    
1. Test simply:
    >>> lazy_matrix_mul([[1, 2], [3, 4]], [[5, 6], [7, 8]])
    matrix([[19, 22],
            [43, 50]])

2. Test 2 invalid matrices:
    >>> lazy_matrix_mul([[1, 2, 3], [3, 4, 5]], [[1, 2], [3, 4]])
    Traceback (most recent call last):
    ...
    ValueError: shapes (2,3) and (2,2) not aligned: 3 (dim 1) != 2 (dim 0)

3. Test 2 valid matrices:
    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    matrix([[19, 22,  5],
            [43, 50, 11],
            [43, 50, 11]])

4. Test singly empty matrix:
    >>> lazy_matrix_mul([], [[3]])
    Traceback (most recent call last):
    ...  
    ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)

5. Test singly empty matrix:
    >>> lazy_matrix_mul([[]], [[3]])
    Traceback (most recent call last):
    ...
    ValueError: shapes (1,0) and (1,1) not aligned: 0 (dim 1) != 1 (dim 0)

6. Test simple case:
    >>> lazy_matrix_mul([[3]], [[9]])
    matrix([[27]])

7. Test string arg:
    >>> lazy_matrix_mul("foo", [[9]])
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead

8. Test string arg:
    >>> lazy_matrix_mul([[8]], "bar")
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead

9. Test int list:
    >>> lazy_matrix_mul([1, 3, 4], [[9]])
    Traceback (most recent call last):
    ...
    TypeError: object of type 'int' has no len()

10. Test int list 2:
SHOULD BE ERROR
    >>> lazy_matrix_mul([[8]], [2, 4, 6])
    Traceback (most recent call last):
    ...
    TypeError: object of type 'int' has no len()


11. Test notnum list:
    >>> lazy_matrix_mul([["foo"]], [[2, 4, 6]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum

12. Test notnum list 2 :
    >>> lazy_matrix_mul([[2, 4, 6]], [["foo"]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum

13. Test a bad rows:
    >>> lazy_matrix_mul([[1, 2], [3, 4, 5], [3, 4]], [[5, 6, 1], [7, 8, 2]])
    Traceback (most recent call last):
    ...
    ValueError: setting an array element with a sequence.

14. Test b bad rows:
    >>> lazy_matrix_mul([[1, 2], [3, 4], [3, 4]], [[5, 6, 1], [7, 2]])
    Traceback (most recent call last):
    ...
    ValueError: setting an array element with a sequence.

15. Test missing 1 arg:
    >>> lazy_matrix_mul([[3]])
    Traceback (most recent call last):
    ...
    TypeError: lazy_matrix_mul() missing 1 required positional argument: 'm_b'

16. Test missing 2 args:
    >>> lazy_matrix_mul()
    Traceback (most recent call last):
    ...
    TypeError: lazy_matrix_mul() missing 2 required positional arguments: 'm_a' and 'm_b'

17. String element:
    >>> lazy_matrix_mul([[5, '6'], [7, 8]], [[5, 6], [7, 8]])
    Traceback (most recent call last):
    ...
    TypeError: invalid data type for einsum


18. Test string arg:
    >>> lazy_matrix_mul("foo", [[9]])
    Traceback (most recent call last):
    ...
    TypeError: Scalar operands are not allowed, use '*' instead
